File: Codes/TSP_Serial.py
import time
import math

# Perform BFS for Hamiltonian cycle
def bfs(node, arr, visited, n):
    start_node = node
    visited_count = 1
    visited[node] = True
    cost = 0
    path = [node]

    while visited_count < n:
        min_cost = math.inf
        next_node = -1

        for i in range(n):
            if not visited[i] and arr[start_node][i] < min_cost:
                min_cost = arr[start_node][i]
                next_node = i

        if next_node == -1:
            break

        visited[next_node] = True
        visited_count += 1
        cost += min_cost
        path.append(next_node)
        start_node = next_node

    cost += arr[start_node][node]
    path.append(node)

    return path, cost

# Update metrics file
def update_metrics(num_vertices, bfs_time, language):
    with open("Metrics.txt", "a") as metrics_out:
        metrics_out.write(f"{num_vertices}\t{bfs_time:.10f}\t{language}\n")

def main(filename):
    try:
        with open(filename, "r") as infile:
            n = int(infile.readline().strip())
            arr = [list(map(int, infile.readline().split())) for _ in range(n)]
    except FileNotFoundError:
        print(f"Error: File '{filename}' not found.")
        return

    paths, costs = [], []
    start_time = time.time()
    for i in range(n):
        visited = [False] * n
        path, cost = bfs(i, arr, visited, n)
        paths.append(path)
        costs.append(cost)
    end_time = time.time()

    bfs_time = end_time - start_time

    with open("Results.txt", "w") as outfile:
        for i in range(n):
            outfile.write(f"Vertex {i}:\n")
            outfile.write(f"Path: {'->'.join(map(str, paths[i]))}\n")
            outfile.write(f"Cost: {costs[i]}\n\n")

    update_metrics(n, bfs_time, "P")
    print(f"BFS Time: {bfs_time:.10f} seconds.")

if __name__ == "__main__":
    import sys
    if len(sys.argv) < 2:
        print("Error: No input file provided.")
    else:
        main(sys.argv[1])

File: Codes/TSP_Serial.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <climits>
#include <ctime>
#include <iomanip>
#include <sstream>

using namespace std;

// Function to perform BFS and calculate Hamiltonian cycle for a specific start node
void BFS(int node, const vector<vector<int>> &arr, vector<int> &visited, int n, string &path, int &cost)
{
  int start_node = node;
  int min_cost;
  int next_node;
  int visited_count = 0;

  visited[start_node] = 1; // Mark starting node as visited
  visited_count++;
  cost = 0;

  stringstream ss;
  ss << node;

  while (visited_count < n)
  {
    min_cost = INT_MAX;
    next_node = -1;

    for (int i = 0; i < n; i++)
    {
      if (!visited[i] && arr[start_node][i] < min_cost)
      {
        min_cost = arr[start_node][i];
        next_node = i;
      }
    }

    if (next_node == -1)
      break;

    visited[next_node] = 1;
    visited_count++;
    cost += min_cost;

    ss << "->" << next_node;
    start_node = next_node;
  }

  cost += arr[start_node][node];
  ss << "->" << node;

  path = ss.str();
}

// Function to update metrics file directly
void updateMetricsFile(const string &filename, int n, double bfs_time, const string &language)
{
  ofstream outfile(filename, ios::app);
  outfile << n << "\t" << fixed << setprecision(10) << bfs_time << "\t" << language << endl;
  outfile.close();
}

int main(int argc, char *argv[])
{
  if (argc < 2)
  {
    cerr << "Usage: " << argv[0] << " <input_filename>" << endl;
    return 1;
  }

  string filename = argv[1];
  ifstream infile(filename);
  if (!infile.is_open())
  {
    cerr << "Error: Unable to open file " << filename << endl;
    return 1;
  }

  int n;
  infile >> n;

  vector<vector<int>> arr(n, vector<int>(n, 0));
  for (int i = 0; i < n; i++)
    for (int j = 0; j < n; j++)
      infile >> arr[i][j];

  infile.close();

  vector<string> paths(n);
  vector<int> costs(n);

  clock_t bfs_start = clock();
  for (int i = 0; i < n; i++)
  {
    vector<int> visited(n, 0);
    BFS(i, arr, visited, n, paths[i], costs[i]);
  }
  clock_t bfs_end = clock();

  double bfs_time = double(bfs_end - bfs_start) / CLOCKS_PER_SEC;

  ofstream outfile("Results.txt");
  for (int i = 0; i < n; i++)
  {
    outfile << "Vertex " << i << ":\n";
    outfile << "Path: " << paths[i] << "\n";
    outfile << "Cost: " << costs[i] << "\n\n";
  }
  outfile.close();

  updateMetricsFile("Metrics.txt", n, bfs_time, "C");

  cout << "BFS Time: " << fixed << setprecision(10) << bfs_time << " seconds." << endl;

  return 0;
}